packages/react-storage/src/components/StorageBrowser/views/hooks/useUploadView.tsx

```diff
@@ -0,0 +1,114 @@
+import React from 'react';
+import { useGetActionInput } from '../../providers/configuration';
+import { useStore } from '../../providers/store';
+import { Task, useProcessTasks } from '../../tasks';
+import { DEFAULT_OVERWRITE_PROTECTION } from '../LocationActionView/constants';
+import { uploadHandler } from '../../actions';
+import { getTaskCounts } from '../../controls/getTaskCounts';
+import { isFunction, isUndefined } from '@aws-amplify/ui';
+
+interface UseUploadView {
+  tasks: Task<File>[];
+  isStartDisabled: boolean;
+  isCancelDisabled: boolean;
+  isOverwriteDisabled: boolean;
+  isSelectFilesDisabled: boolean;
+  overwriteSelection: boolean;
+  onDropFiles: (files?: File[]) => void;
+  onExit: () => void;
+  onProcessStart: () => void;
+  onProcessCancel: () => void;
+  onSelectFiles: (type?: 'FILE' | 'FOLDER') => void;
+  onToggleOverwrite: () => void;
+}
+
+export const useUploadView = ({
+  onClose,
+}: {
+  onClose?: () => void;
+}): UseUploadView => {
+  const getInput = useGetActionInput();
+  const [{ files, history }, dispatchStoreAction] = useStore();
+  const { prefix } = history?.current ?? {};
+
+  const hasInvalidPrefix = isUndefined(prefix);
+  const [overwriteSelection, setOverwriteSelection] = React.useState(
+    DEFAULT_OVERWRITE_PROTECTION
+  );
+
+  const [tasks, handleProcess] = useProcessTasks(uploadHandler, files, {
+    concurrency: 4,
+  });
+
+  const taskCounts = React.useMemo(() => getTaskCounts(tasks), [tasks]);
+
+  const hasStarted = !!taskCounts.PENDING;
+  const hasCompleted =
+    !!taskCounts.TOTAL &&
+    taskCounts.CANCELED + taskCounts.COMPLETE + taskCounts.FAILED ===
+      taskCounts.TOTAL;
+
+  const isCancelDisabled = !taskCounts.TOTAL || !hasStarted || hasCompleted;
+  const isStartDisabled = !taskCounts.TOTAL || hasStarted || hasCompleted;
+  const isOverwriteDisabled = hasStarted || hasCompleted;
+  const isSelectFilesDisabled = hasStarted || hasCompleted;
+
+  const onDropFiles = React.useCallback(
+    (files?: File[]) => {
+      if (files) {
+        dispatchStoreAction({ type: 'ADD_FILE_ITEMS', files });
+      }
+    },
+    [dispatchStoreAction]
+  );
+
+  const onSelectFiles = React.useCallback(
+    (type?: 'FILE' | 'FOLDER') => {
+      dispatchStoreAction({ type: 'SELECT_FILES', selectionType: type });
+    },
+    [dispatchStoreAction]
+  );
+
+  const onProcessStart = React.useCallback(() => {
+    if (hasInvalidPrefix) return;
+
+    handleProcess({
+      config: getInput(),
+      prefix,
+      options: { preventOverwrite: overwriteSelection },
+    });
+  }, [handleProcess, prefix, overwriteSelection, getInput, hasInvalidPrefix]);
+
+  const onProcessCancel = React.useCallback(() => {
+    tasks.forEach((task) => task.cancel?.());
+  }, [tasks]);
+
+  const onExit = React.useCallback(() => {
+    if (isFunction(onClose)) onClose?.();
+    // clear tasks state
+    tasks.forEach(({ remove }) => remove());
```

Comment:
Curious if this is necessary since we're unmounting the component. Is this just to prevent a mistake where tasks keep running?

Reply:
The hook might be used outside the context of the component, realistically all of our `onExit` callbacks should clean up state